File: 0802-find-eventual-safe-states/0802-find-eventual-safe-states.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n=graph.size();
        vector<vector<int>>adj(n);
        vector<int>indeg(n,0);
        for(int i=0;i<n;i++){
            for(auto it:graph[i]){
                adj[it].push_back(i);
                indeg[i]++;
            }
        }
        queue<int>q;
        vector<int>ans;
        for(int i=0;i<n;i++){
            if(indeg[i]==0){
                q.push(i);
            }
        }

        while(!q.empty()){
            auto top=q.front();
            q.pop();
            ans.push_back(top);
            for(auto it:adj[top]){
                indeg[it]--;
                if(indeg[it]==0){
                    q.push(it);
                }
            }
        }
        sort(ans.begin(),ans.end());
        return ans;
    }
};

File: 0802-find-eventual-safe-states/0802-find-eventual-safe-states.java
class Solution {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        int n=graph.length;
        int[] indeg=new int[n];

        List<List<Integer>> rev=new ArrayList<>();
        for(int i=0;i<n;i++){
            rev.add(new ArrayList<>());
        }

        for(int i=0;i<n;i++){
            for(int nei:graph[i]){
                rev.get(nei).add(i);
                indeg[i]++;
            }
        }

        Queue<Integer> q=new LinkedList<>();

        for(int i=0;i<n;i++){
            if(indeg[i]==0){
                q.offer(i);
            }
        }

        boolean[] safe = new boolean[n];
        List<Integer> ans=new ArrayList<>();

        int idx=0;

        while(!q.isEmpty()){
            int node=q.poll();
            safe[node]=true;

            for(int nei:rev.get(node)){
                indeg[nei]--;
                if(indeg[nei]==0){
                    q.offer(nei);
                }
            }
        }
        for (int i = 0; i < n; i++) {
            if (safe[i]) ans.add(i);
        }
        return ans;
    }
}